core/src/main/scala/akka/persistence/jdbc/state/DurableStateQueries.scala

```diff
@@ -0,0 +1,19 @@
+package akka.persistence.jdbc.state
+
+import slick.jdbc.JdbcProfile
+import akka.persistence.jdbc.config.DurableStateTableConfiguration
+
+class DurableStateQueries(val profile: JdbcProfile, override val durableStateTableCfg: DurableStateTableConfiguration)
+    extends DurableStateTables {
+  import profile.api._
+
+  def _selectByPersistenceId(persistenceId: Rep[String]) =
+    durableStateTable.filter(_.persistenceId === persistenceId)
+
+  def _upsertDurableState(row: DurableStateTables.DurableStateRow) =
+    durableStateTable.insertOrUpdate(row)
```

Comment:
will this automatically verify the expected seqNr?

Reply:
What this does is if the record with the primary key `persistenceId` does not exist, it inserts. Otherwise updates. And it does this _transactionally_. Do we need any other validation ? Because if we use an if statement here the total operation will not be transactional. As per the spanner implementation ..

```scala
if (seqNr == 1) {
  "insert... "
} else {
  val updatedRows = "update foo ... where seq_nr = ${seqNr - 1}" 
  if (updatedRows != 1)
    throw boom
}
``` 

2 writers may try to execute the above code. Both pass the if check and tries to `insert`. Since the primary key is the `persistenceId` both will go through and we will have 2 records with `seqNr` = 1. Please suggest what logic should we follow here. I implemented the basic `upsert` support that Slick offers if the underlying database implements an upsert.

Reply:
I don't know what slick provides but in the end it would be good if the update could include the where condition on the expected seqNr. `"update foo ... where seq_nr = ${seqNr - 1}"`.

This kind of update condition is often called "optimistic" locking if you can search for if Slick has any built in support for that.

I understand the problem with seqNr 1, but if we can't have both checks I prefer to have it on the updates. It's a best effort check to detect concurrent writers, and if an application has that problem it's more likely that we detect it for updates (seqNr > 1) than for the first insert.

Reply:
@patriknw a few more questions ..

* For the update statement we need to check for the persistenceId match in the where clause as well - right ? e.g. `update state ... where persistence_id = $persistenceId and sequence_number = ${seqNr - 1}`

* Also if we make a composite primary key `persistence_id + sequence_number` then the possibility of inserting 2 records with the same sequence number is eliminated at the database level. So if the `insert` is executed concurrently for the same `persistence_id + sequence_number` one of them will fail - does this sound ok ?

* And in the update above what happens if the number of updated rows is `0` i.e. an incorrect sequence number is provided by the user - should we throw ? And note that there cannot be multiple records being updated as we cannot have multiple instances of the same `persistence_id + sequence_number` combination.

Reply:
> persistenceId match in the where clause as well

absolutely, that's the primary key

> composite primary key persistence_id + sequence_number

but then it will not be updates the existing row if it's a new primary key for each seqNr increment

> an incorrect sequence number is provided by the user - should we throw ?

yes, that should throw, look at Spanner implementation

> note that there cannot be multiple records being updated

right, the persistenceId is the primary key and there can be 0 or 1 rows for a given persistenceId

Reply:
Surprisingly there seems to be no support for optimistic locking in Slick https://github.com/slick/slick/issues/378

We could implement it with [plain sql](https://scala-slick.org/doc/3.3.3/sql.html) but let's skip this feature of checking the sequence number altogether. We want to move away from Slick anyway and then we can implement it with that library.

Reply:
As far as I can tell, Slick does not provide any out-of-the-box optimistic locking, but we can implement if with a where clause in the update. 

> For the update statement we need to check for the persistenceId match in the where clause as well - right ? e.g. update state ... where persistence_id = $persistenceId and sequence_number = ${seqNr - 1}

Yes, in that case we need both. 

> Also if we make a composite primary key persistence_id + sequence_number then the possibility of inserting 2 records with the same sequence number is eliminated at the database level. So if the insert is executed concurrently for the same persistence_id + sequence_number one of them will fail - does this sound ok ?

I think this is ok, we let the DB take care of that. 

> And in the update above what happens if the number of updated rows is 0 i.e. an incorrect sequence number is provided by the user - should we throw ? And note that there cannot be multiple records being updated as we cannot have multiple instances of the same persistence_id + sequence_number combination.

The user in our case is the actor that is tracking the `seq_num`, not the real user. 
We can run into a situation where two actors are in memory, pointing to the same state and holding the same seq_num, for example on a badly configured cluster or a cluster that has a split brain. 

In that case, the optimistic locking will be the last change to avoid it. One of the actors will crash and that's good.